client/js/analysis/gdpr.js

```diff
@@ -0,0 +1,199 @@
+// @ts-check
+/**
+ * @typedef {object} RawGDPRRecord
+ * @property {string} endTime Date and time of when the stream ended in UTC
+ * format.
+ * @property {string} artistName Name of "creator" for each stream (e.g. the
+ * artist name if a music track).
+ * @property {string} trackName Name of items listened to or watched (e.g. title
+ * of music track or name of video).
+ * @property {number} msPlayed How many mili-seconds the
+ * track was listened.
+ */
+
+/**
+ * @typedef {object} GDPRRecord
+ * @property {Date} endTime Date and time of when the stream ended in UTC
+ * format.
+ * @property {string} artistName Name of "creator" for each stream (e.g. the
+ * artist name if a music track).
+ * @property {string} trackName Name of items listened to or watched (e.g. title
+ * of music track or name of video).
+ * @property {number} msPlayed How many mili-seconds the
+ * track was listened.
+ */
+
+// global variables are not available in ES modules
+const {zip} = window;
+
+/**
+ * Takes a blob containing zipped GDPR data and outputs a list of GDPR streaming
+ * records.
+ *
+ * @param {Blob} data The zipped GDPR data.
+ * @returns {Promise<GDPRRecord[]>} A list of GDPR streaming records.
+ */
+export async function getStreamingData(data) {
+  const streamingDataPattern = /StreamingHistory(\d+)\.json$/i;
+
+  return new Promise((resolve, reject) => {
+    zip.createReader(
```

Comment:
The functions in this reader are quite difficult to tease out given the heavy nesting of inlined-functions. You should instead split these functions out into separate functions that you call:
```
export async function getStreamingData(data) {
  ...
  const fileDataToUtcEntry = (fileData) => {
    ...
  };
 
  const readFiles = (files) => {
    ...
    const contentPromises = entriesWithIndices.map(...).map(fileDataToUtcEntry);
    Promise.all(contentPromises).then(entries => resolve(entries));
  }

  zip.createReader(new zip.BlobReader(data), (reader) => reader.getEntries(readFiles));
```